Src/Support/Google.Apis.Auth/OAuth2/AwsExternalAccountCredential.AwsSecurityCredential.cs

```diff
@@ -0,0 +1,144 @@
+            /// </summary>
+            internal static AwsSecurityCredentials MaybeFromEnvVars()
+            {
+                if (AccessKeyIdEnvVar.GetNonEmptyEnvVarValue() is string accessKeyId
+                    && SecretAccessKeyEnvVar.GetNonEmptyEnvVarValue() is string secretAccessKey)
+                {
+                    return new AwsSecurityCredentials(accessKeyId, secretAccessKey, TokenEnvVar.GetNonEmptyEnvVarValue());
+                }
+
+                return null;
+            }
+
+            /// <summary>
+            /// Attempts to fetch the security credentials from the metadata server.
+            /// Returns null if the credential URL is null or empty.
+            /// </summary>
+            internal static async Task<AwsSecurityCredentials> MaybeFromMetadataAsync(AwsMetadataServerClient metadataClient, string credentialUrl)
+            {
+                // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+                if (!string.IsNullOrEmpty(credentialUrl)
+                    // Build the actual credential endpoint.
+                    && await BuildSecurityCredentialsEndpointAsync().ConfigureAwait(false) is string credentialEndpoint
+                    // Fetch the credential JSON
+                    && await metadataClient.FetchMetadataAsync(credentialEndpoint).ConfigureAwait(false) is string credentialJson
+                    && credentialJson != "")
+                {
+                    // Deserialize the credentials
+                    var deserializedCredentials = NewtonsoftJsonSerializer.Instance.Deserialize<AwsSecurityCredentialsResponse>(credentialJson);
+
+                    if (deserializedCredentials.IsSuccess
+                        && !string.IsNullOrEmpty(deserializedCredentials.AccessKeyId)
+                        && !string.IsNullOrEmpty(deserializedCredentials.SecretAccessKey))
+                    {
+                        return new AwsSecurityCredentials(
+                            deserializedCredentials.AccessKeyId, deserializedCredentials.SecretAccessKey,
+                            deserializedCredentials.Token == "" ? null : deserializedCredentials.Token);
+                    }
+
+                    throw new InvalidOperationException(
+                        $"The request to the metadata server security credentials endpoint {credentialEndpoint} was not succesful." +
+                        $"the response was {credentialJson}");
+                }
+
+                return null;
+
+                async Task<string> BuildSecurityCredentialsEndpointAsync()
+                {
+                    // https://github.com/aws/aws-sdk-java/blob/ccfd63c097874e3a1e9ffda7bf171769b6bd3b21/aws-java-sdk-core/src/main/java/com/amazonaws/auth/InstanceMetadataServiceCredentialsFetcher.java#L55
+                    // Fetch the list of roles attached to the instance from the credential URL as given.
```

Comment:
If it's a *list* of roles, is that line-separated, comma-separated, something else? Which role are we going to pick?
Even if the code is doing the right thing at the moment, we should add comments to explain.

(My guess is "if there are multiple roles, they're on different lines, and we just pick the first of them")